src/app/SubscriptionResumptionStorage.h

```diff
@@ -0,0 +1,158 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the interface to store subscription information.
+ */
+
+#pragma once
+
+#include <app/ReadClient.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/support/CommonIterator.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The SubscriptionResumptionStorage interface is used to persist subscriptions when they are established.
+ */
+class SubscriptionResumptionStorage
+{
+public:
+    // Structs to hold path param values as is_trivial struct
+    struct AttributePathParamsValues
+    {
+        ClusterId mClusterId;
+        AttributeId mAttributeId;
+        EndpointId mEndpointId;
+        void SetValues(const AttributePathParams & params)
+        {
+            mEndpointId  = params.mEndpointId;
+            mClusterId   = params.mClusterId;
+            mAttributeId = params.mAttributeId;
+        }
+        AttributePathParams GetParams() { return AttributePathParams(mEndpointId, mClusterId, mAttributeId); }
+    };
+    struct EventPathParamsValues
+    {
+        ClusterId mClusterId;
+        EventId mEventId;
+        EndpointId mEndpointId;
+        bool mIsUrgentEvent;
+        void SetValues(const EventPathParams & params)
+        {
+            mEndpointId    = params.mEndpointId;
+            mClusterId     = params.mClusterId;
+            mEventId       = params.mEventId;
+            mIsUrgentEvent = params.mIsUrgentEvent;
+        }
+        EventPathParams GetParams() { return EventPathParams(mEndpointId, mClusterId, mEventId, mIsUrgentEvent); }
+    };
+
+    /**
+     * Struct to hold information about subscriptions
+     */
+    struct SubscriptionInfo
+    {
+        NodeId mNodeId;
+        FabricIndex mFabricIndex;
+        SubscriptionId mSubscriptionId;
+        uint16_t mMinInterval;
+        uint16_t mMaxInterval;
+        bool mFabricFiltered;
+        Platform::ScopedMemoryBufferWithSize<AttributePathParamsValues> mAttributePaths;
+        Platform::ScopedMemoryBufferWithSize<EventPathParamsValues> mEventPaths;
+        CHIP_ERROR SetAttributePaths(const ObjectList<AttributePathParams> * pAttributePathList)
+        {
+            mAttributePaths.Free();
+            const ObjectList<AttributePathParams> * attributePath = pAttributePathList;
+            size_t attributePathCount                             = 0;
+            while (attributePath)
+            {
+                attributePathCount++;
+                attributePath = attributePath->mpNext;
+            }
+            ReturnErrorCodeIf((attributePathCount * sizeof(AttributePathParamsValues)) > UINT16_MAX, CHIP_ERROR_NO_MEMORY);
```

Comment:
Why do we need this check?  I would have thought Calloc() handles it for us....  If not, it should, and this needs a followup issue.